icu4c/source/i18n/unitsrouter.cpp

```diff
@@ -0,0 +1,74 @@
+// © 2020 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html
+
+#include "unicode/utypes.h"
+
+#if !UCONFIG_NO_FORMATTING
+
+#include <utility>
+
+#include "cmemory.h"
+#include "cstring.h"
+#include "number_decimalquantity.h"
+#include "resource.h"
+#include "unitconverter.h" // for extractCompoundBaseUnit
+#include "unitsdata.h"     // for getUnitCategory
+#include "unitsrouter.h"
+#include "uresimp.h"
+
+U_NAMESPACE_BEGIN
+
+UnitsRouter::UnitsRouter(MeasureUnit inputUnit, StringPiece region, StringPiece usage,
+                         UErrorCode &status) {
+    // TODO: do we want to pass in ConversionRates and UnitPreferences instead
+    // of loading in each UnitsRouter instance? (Or make global?)
+    ConversionRates conversionRates(status);
+    UnitPreferences prefs(status);
+
+    MeasureUnit baseUnit = extractCompoundBaseUnit(inputUnit, conversionRates, status);
+    CharString category = getUnitCategory(baseUnit.getIdentifier(), status);
+
+    // TODO: deal correctly with StringPiece / null-terminated string incompatibility...
+    const UnitPreference *const *unitPreferences;
+    int32_t preferencesCount;
+    prefs.getPreferencesFor(category.data(), usage.data(), region.data(), unitPreferences,
+                            preferencesCount, status);
+
+    for (int i = 0; i < preferencesCount; ++i) {
+        const auto &preference = *unitPreferences[i];
+        MeasureUnit complexTargetUnit = MeasureUnit::forIdentifier(preference.unit.data(), status);
+
+        if (U_FAILURE(status)) {
+          return;
+        }
+        
+        converterPreferences_.emplaceBack(inputUnit, complexTargetUnit,
+                                          preference.geq, conversionRates,
+                                          status);
+        if (U_FAILURE(status)) {
+            fprintf(
+                stderr,
+                "FAILED: converterPreferences_.emplaceBack(<%s>, <%s>, %f, conversionRates, status)\n",
+                inputUnit.getIdentifier(), complexTargetUnit.getIdentifier(), preference.geq);
+            return;
+        }
+    }
+}
+
+MaybeStackVector<Measure> UnitsRouter::route(double quantity, UErrorCode &status) {
+    for (int i = 0, n = converterPreferences_.length() - 1; i < n; i++) {
```

Comment:
Triggers a warning:

    warning: control may reach end of non-void function [-Wreturn-type]

The end condition could be removed, to be explicit about the "return" being the only way out of the loop - e.g.:

    for (int i = 0, n = converterPreferences_.length() - 1;; i++) {

Reply:
Done!